### PR TITLE
Defines for Expansion version number

### PR DESCRIPTION
**Pull request by mrgriffin**
Currently unused, except to display the Expansion version on the debug menu:
image

Marked as a draft, because I think it might be worth discussing what we want to write in the version number field. "v1.6.1" is misleading, because `master` includes more changes than that. But "v1.6.2" is also misleading because `master` is missing some changes that will be in that.

I'm not sure if this could have any other uses. You could imagine, e.g. using it as part of the link battle process so that whoever has the latest version of Expansion becomes the host... but tbh we'd probably want to pick the latest version of the hack rather than Expansion.

### CODE REVIEW

**Comment by Bassoonian** [comment]
I'm not sure if this is possible, but perhaps through preproc magic we could include the hash of the latest commit?

**Review by AlexOn1ine — COMMENTED** [review]
Maybe just v1.6? Since this would be the least misleading but obviously omits some information. Looks great otherwise!

**Comment by mrgriffin** [comment]
> I'm not sure if this is possible, but perhaps through preproc magic we could include the hash of the latest commit? Or alternatively the hash of something else (an elf file?)

I think it would be reasonably easy to inject a hash in via a rule in the `Makefile`, but because Expansion is used in other people's projects I'm not sure if we have easy access to a hash that only updates when the Expansion updates :thinking: 

> I think maybe adding a `#define EXPANSION_VERSION_UNRELEASED TRUE` could be helpful here, and changing this during the release workflow.

Hmm... This sounds pretty doable. Would it get annoying for you to have to make a commit with the numbers updated and that defined to `TRUE`, tag that, then immediately create another commit with it set back to `FALSE`?

**Comment by AsparagusEduardo** [comment]
> I think it would be reasonably easy to inject a hash in via a rule in the `Makefile`, but because Expansion is used in other people's projects I'm not sure if we have easy access to a hash that only updates when the Expansion updates 🤔

Yeah, that'd be hard to keep track of.

> Hmm... This sounds pretty doable. Would it get annoying for you to have to make a commit with the numbers updated and that defined to `TRUE`, tag that, then immediately create another commit with it set back to `FALSE`?

Not annoying, but I might forget like how I sometimes forget to update the issue templates ^^;

**Comment by AsparagusEduardo** [comment]
Would you say this is ready for merge?
Even though it's currently unused, I feel we should've had this for a while now ^^;

**Comment by mrgriffin** [comment]
Yeah I think it'd be reasonable to merge as-is. I suppose the one other thing we might want to do, possibly part of this idk, is put the version number in some sort of `rom_header_rhh.c` kinda file, so that external tools are able to extract it out of the compiled ROM?

**Comment by AsparagusEduardo** [comment]
> Yeah I think it'd be reasonable to merge as-is. I suppose the one other thing we might want to do, possibly part of this idk, is put the version number in some sort of `rom_header_rhh.c` kinda file, so that external tools are able to extract it out of the compiled ROM?

Oh yeah, that might be useful.

**Comment by AsparagusEduardo** [comment]
Thinking about it, maybe this could be a thing for 1.7, due to being a bit awkward for versions 1.6.0 and 1.6.1 to not have this, but 1.6.2 doing so.

**Comment by mrgriffin** [comment]
> > I suppose the one other thing we might want to do is put the version number in some sort of `rom_header_rhh.c` kinda file, so that external tools are able to extract it out of the compiled ROM?
> 
> Oh yeah, that might be useful.

I've pushed a very simple version of this that just contains some magic bytes (to help locate the header if, e.g. `GFRomHeader` grows/shrinks for whatever downstream reason), plus the Expansion version number.

Is there anything else worth adding?

If not, ready for review.

**Comment by DizzyEggg** [comment]
CI failed 👀

**Comment by mrgriffin** [comment]
> CI failed 👀

I'm an idiot who forgets to `git add` new files :sweat_smile:

File: src/rom_header_rhh.c [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+// Similar to the GF ROM header, this struct allows external programs to
+// detect details about Expansion.
+// For this structure to be useful we have to maintain backwards binary
+// compatibility. This means that we should only ever append data to the
+// end. If there are any structs as members then those structs should
+// not be modified after being introduced.
+struct RHHRomHeader
+{
+    char rhh_magic[4]; // 'PORY'. Useful to locate the header if it shifts.
```
Comment (AsparagusEduardo) [inline src/rom_header_rhh.c:12]: I'm thinking we need a longer word, seeing that only 4 characters might give false-positives due to how likely it is to find 4 bytes with that combination.

File: src/rom_header_rhh.c [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    u8 expansionVersionMajor;
+    u8 expansionVersionMinor;
+    u8 expansionVersionPatch;
+    bool8 expansionVersionTagged;
```
Comment (AsparagusEduardo) [inline src/rom_header_rhh.c:16]:
```suggestion
    bool8 expansionVersionTagged:1;
    u8 padding:7;
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/rom_header_rhh.c (hunk 1 above)
Comment (AsparagusEduardo) [reply src/rom_header_rhh.c:12]: Maybe `'RHHEXP'`?

File: src/rom_header_rhh.c (hunk 2 above)
Comment (DizzyEggg) [reply src/rom_header_rhh.c:16]: What's the reasoning behind it actually?

Comment (AsparagusEduardo) [reply src/rom_header_rhh.c:16]: Behind `expansionVersionTagged`? or the padding?

Comment (DizzyEggg) [reply src/rom_header_rhh.c:16]: the padding

Comment (AsparagusEduardo) [reply src/rom_header_rhh.c:16]: I figured that since we would only need a single bit for this, might as well just use what we need?

**Comment by AsparagusEduardo** [comment]
With this new name, it makes we wonder if we should get ahead and reserve a full u32 for flags or not

**Comment by mrgriffin** [comment]
I think we can always add more flag fields in the future if we find that we need more than 8 :)

**Review by AsparagusEduardo — APPROVED** [review]

**Comment by AsparagusEduardo** [comment]
Lgtm, though thinking about it I'm wondering about how we're going to organize the versions.
Would the version numbers reflect the next version to be released or the latest one released?

**Comment by mrgriffin** [comment]
Repointed at upcoming. Please squash :)

Hmm. I'm not sure tbh! I suppose the next version might make sense, which means that `upcoming` and `master` would have different version numbers. But I don't feel strongly, so I can change it if you want. lmk :)

**Comment by AsparagusEduardo** [comment]
> Repointed at upcoming. Please squash :)
> 
> Hmm. I'm not sure tbh! I suppose the next version might make sense, which means that `upcoming` and `master` would have different version numbers. But I don't feel strongly, so I can change it if you want. lmk :)

I think having the two branches be different version numbers is fine, since upcoming is supposed to be the "minor" version increase (based on semver).

After the 1.7.0 release in master, I could change the tagged version to false and bump the version to 1.7.1-unreleased, while upcoming would be changed to 1.8.0-unreleased.

Does that sound good to you?

**Comment by mrgriffin** [comment]
Sounds good to me! :)

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**